File: servidor-java/src/link/Hamming.java
package link;

import java.util.Arrays;


public final class Hamming {
    private static final int DATA_BLOCK_SIZE = 8;
    private static final int CODEWORD_BLOCK_SIZE = 12;

    private static final int[] PARITY_POSITIONS = {
        1, 2, 4, 8
    };

    private static final int[] DATA_POSITIONS = {
        3, 5, 6, 7, 9, 10, 11, 12
    };

    private Hamming() {
    }

    private static int[] getCheckedPositions(
        int parityPosition
    ) {
        if (parityPosition == 1) {
            return new int[] {1, 3, 5, 7, 9, 11};
        }

        if (parityPosition == 2) {
            return new int[] {2, 3, 6, 7, 10, 11};
        }

        if (parityPosition == 4) {
            return new int[] {4, 5, 6, 7, 12};
        }

        if (parityPosition == 8) {
            return new int[] {8, 9, 10, 11, 12};
        }

        throw new IllegalArgumentException(
            "Posición de paridad desconocida: "
            + parityPosition
        );
    }

    private static void validateBits(
        String bits,
        String fieldName
    ) {
        if (bits == null || bits.isEmpty()) {
            throw new IllegalArgumentException(
                fieldName + " no puede estar vacío."
            );
        }

        for (
            int currentBitPosition = 0;
            currentBitPosition < bits.length();
            currentBitPosition++
        ) {
            char currentBit = bits.charAt(currentBitPosition);

            if (currentBit != '0' && currentBit != '1') {
                throw new IllegalArgumentException(
                    fieldName + " solo puede contener 0 y 1."
                );
            }
        }
    }

    private static String encodeBlock(
        String dataBlockBits
    ) {
        char[] codewordPositions =
            new char[CODEWORD_BLOCK_SIZE + 1];
        Arrays.fill(codewordPositions, '0');

        for (
            int dataIndex = 0;
            dataIndex < DATA_BLOCK_SIZE;
            dataIndex++
        ) {
            int codewordPosition = DATA_POSITIONS[dataIndex];
            codewordPositions[codewordPosition] =
                dataBlockBits.charAt(dataIndex);
        }

        for (int parityPosition : PARITY_POSITIONS) {
            int[] checkedPositions =
                getCheckedPositions(parityPosition);
            int onesCount = 0;

            for (int checkedPosition : checkedPositions) {
                if (codewordPositions[checkedPosition] == '1') {
                    onesCount++;
                }
            }

            if (onesCount % 2 == 0) {
                codewordPositions[parityPosition] = '0';
            } else {
                codewordPositions[parityPosition] = '1';
            }
        }

        return new String(
            codewordPositions,
            1,
            CODEWORD_BLOCK_SIZE
        );
    }

    public static String encode(String dataBits) {
        validateBits(dataBits, "Los datos");

        if (dataBits.length() % DATA_BLOCK_SIZE != 0) {
            throw new IllegalArgumentException(
                "La longitud de los datos debe ser "
                + "múltiplo de 8."
            );
        }

        StringBuilder encodedBlocks = new StringBuilder();

        for (
            int blockStartPosition = 0;
            blockStartPosition < dataBits.length();
            blockStartPosition += DATA_BLOCK_SIZE
        ) {
            int blockEndPosition =
                blockStartPosition + DATA_BLOCK_SIZE;
            String dataBlockBits = dataBits.substring(
                blockStartPosition,
                blockEndPosition
            );
            encodedBlocks.append(encodeBlock(dataBlockBits));
        }

        return encodedBlocks.toString();
    }

    private static int calculateErrorPosition(
        String codewordBlockBits
    ) {
        char[] codewordPositions =
            new char[CODEWORD_BLOCK_SIZE + 1];

        for (
            int codewordIndex = 0;
            codewordIndex < CODEWORD_BLOCK_SIZE;
            codewordIndex++
        ) {
            int codewordPosition = codewordIndex + 1;
            codewordPositions[codewordPosition] =
                codewordBlockBits.charAt(codewordIndex);
        }

        int errorPosition = 0;

        for (int parityPosition : PARITY_POSITIONS) {
            int[] checkedPositions =
                getCheckedPositions(parityPosition);
            int onesCount = 0;

            for (int checkedPosition : checkedPositions) {
                if (codewordPositions[checkedPosition] == '1') {
                    onesCount++;
                }
            }

            boolean parityFailed = onesCount % 2 != 0;

            if (parityFailed) {
                errorPosition += parityPosition;
            }
        }

        return errorPosition;
    }

    private static void validateCodeword(
        String codewordBits
    ) {
        validateBits(codewordBits, "La palabra código");

        if (
            codewordBits.length() % CODEWORD_BLOCK_SIZE != 0
        ) {
            throw new IllegalArgumentException(
                "La longitud de la palabra código debe ser "
                + "múltiplo de 12."
            );
        }
    }

    public static boolean verify(String codewordBits) {
        validateCodeword(codewordBits);

        for (
            int blockStartPosition = 0;
            blockStartPosition < codewordBits.length();
            blockStartPosition += CODEWORD_BLOCK_SIZE
        ) {
            int blockEndPosition =
                blockStartPosition + CODEWORD_BLOCK_SIZE;
            String codewordBlockBits = codewordBits.substring(
                blockStartPosition,
                blockEndPosition
            );

            if (calculateErrorPosition(codewordBlockBits) != 0) {
                return false;
            }
        }

        return true;
    }

    public static String correct(String codewordBits) {
        validateCodeword(codewordBits);
        StringBuilder correctedBlocks = new StringBuilder();

        for (
            int blockStartPosition = 0;
            blockStartPosition < codewordBits.length();
            blockStartPosition += CODEWORD_BLOCK_SIZE
        ) {
            int blockEndPosition =
                blockStartPosition + CODEWORD_BLOCK_SIZE;
            String codewordBlockBits = codewordBits.substring(
                blockStartPosition,
                blockEndPosition
            );
            int errorPosition =
                calculateErrorPosition(codewordBlockBits);
            char[] correctedBlockBits =
                codewordBlockBits.toCharArray();

            if (errorPosition > CODEWORD_BLOCK_SIZE) {
                throw new IllegalArgumentException(
                    "Hamming detectó un síndrome fuera "
                    + "del bloque de 12 bits."
                );
            }

            if (errorPosition != 0) {
                int errorIndex = errorPosition - 1;

                if (correctedBlockBits[errorIndex] == '0') {
                    correctedBlockBits[errorIndex] = '1';
                } else {
                    correctedBlockBits[errorIndex] = '0';
                }
            }

            correctedBlocks.append(correctedBlockBits);
        }

        return correctedBlocks.toString();
    }

    public static String recoverData(String codewordBits) {
        String correctedCodewordBits = correct(codewordBits);
        StringBuilder recoveredDataBits = new StringBuilder();

        for (
            int blockStartPosition = 0;
            blockStartPosition < correctedCodewordBits.length();
            blockStartPosition += CODEWORD_BLOCK_SIZE
        ) {
            int blockEndPosition =
                blockStartPosition + CODEWORD_BLOCK_SIZE;
            String correctedBlockBits =
                correctedCodewordBits.substring(
                    blockStartPosition,
                    blockEndPosition
                );

            for (int dataPosition : DATA_POSITIONS) {
                int dataIndex = dataPosition - 1;
                recoveredDataBits.append(
                    correctedBlockBits.charAt(dataIndex)
                );
            }
        }

        return recoveredDataBits.toString();
    }
}

File: cliente-python/src/link/hamming.py
class Hamming:
    DATA_BLOCK_SIZE = 8
    CODEWORD_BLOCK_SIZE = 12

    PARITY_POSITIONS = (1, 2, 4, 8)
    DATA_POSITIONS = (3, 5, 6, 7, 9, 10, 11, 12)

    PARITY_GROUPS = {
        1: (1, 3, 5, 7, 9, 11),
        2: (2, 3, 6, 7, 10, 11),
        4: (4, 5, 6, 7, 12),
        8: (8, 9, 10, 11, 12),
    }

    @staticmethod
    def _validate_bits(bits: str, field_name: str):
        if not bits:
            raise ValueError(
                f"{field_name} no puede estar vacío."
            )

        for bit in bits:
            if bit != "0" and bit != "1":
                raise ValueError(
                    f"{field_name} solo puede contener 0 y 1."
                )

    @staticmethod
    def _encode_block(data_block_bits: str) -> str:
        codeword_positions = ["0"] * (
            Hamming.CODEWORD_BLOCK_SIZE + 1
        )

        for data_index in range(Hamming.DATA_BLOCK_SIZE):
            codeword_position = Hamming.DATA_POSITIONS[
                data_index
            ]
            codeword_positions[codeword_position] = (
                data_block_bits[data_index]
            )

        for parity_position in Hamming.PARITY_POSITIONS:
            checked_positions = Hamming.PARITY_GROUPS[
                parity_position
            ]
            ones_count = 0

            for checked_position in checked_positions:
                if codeword_positions[checked_position] == "1":
                    ones_count += 1

            if ones_count % 2 == 0:
                parity_bit = "0"
            else:
                parity_bit = "1"

            codeword_positions[parity_position] = parity_bit

        return "".join(codeword_positions[1:])

    @staticmethod
    def encode(data_bits: str) -> str:
        Hamming._validate_bits(data_bits, "Los datos")

        if len(data_bits) % Hamming.DATA_BLOCK_SIZE != 0:
            raise ValueError(
                "La longitud de los datos debe ser "
                "múltiplo de 8."
            )

        encoded_blocks = []

        for block_start_position in range(
            0,
            len(data_bits),
            Hamming.DATA_BLOCK_SIZE,
        ):
            block_end_position = (
                block_start_position
                + Hamming.DATA_BLOCK_SIZE
            )
            data_block_bits = data_bits[
                block_start_position:block_end_position
            ]
            encoded_block_bits = Hamming._encode_block(
                data_block_bits
            )
            encoded_blocks.append(encoded_block_bits)

        return "".join(encoded_blocks)

    @staticmethod
    def _calculate_error_position(
        codeword_block_bits: str,
    ) -> int:
        codeword_positions = [""] + list(
            codeword_block_bits
        )
        error_position = 0

        for parity_position in Hamming.PARITY_POSITIONS:
            checked_positions = Hamming.PARITY_GROUPS[
                parity_position
            ]
            ones_count = 0

            for checked_position in checked_positions:
                if codeword_positions[checked_position] == "1":
                    ones_count += 1

            parity_failed = ones_count % 2 != 0

            if parity_failed:
                error_position += parity_position

        return error_position

    @staticmethod
    def _validate_codeword(codeword_bits: str):
        Hamming._validate_bits(
            codeword_bits,
            "La palabra código",
        )

        if (
            len(codeword_bits)
            % Hamming.CODEWORD_BLOCK_SIZE
            != 0
        ):
            raise ValueError(
                "La longitud de la palabra código debe ser "
                "múltiplo de 12."
            )

    @staticmethod
    def verify(codeword_bits: str) -> bool:
        Hamming._validate_codeword(codeword_bits)

        for block_start_position in range(
            0,
            len(codeword_bits),
            Hamming.CODEWORD_BLOCK_SIZE,
        ):
            block_end_position = (
                block_start_position
                + Hamming.CODEWORD_BLOCK_SIZE
            )
            codeword_block_bits = codeword_bits[
                block_start_position:block_end_position
            ]
            error_position = (
                Hamming._calculate_error_position(
                    codeword_block_bits
                )
            )

            if error_position != 0:
                return False

        return True

    @staticmethod
    def correct(codeword_bits: str) -> str:
        Hamming._validate_codeword(codeword_bits)
        corrected_blocks = []

        for block_start_position in range(
            0,
            len(codeword_bits),
            Hamming.CODEWORD_BLOCK_SIZE,
        ):
            block_end_position = (
                block_start_position
                + Hamming.CODEWORD_BLOCK_SIZE
            )
            codeword_block_bits = codeword_bits[
                block_start_position:block_end_position
            ]
            error_position = (
                Hamming._calculate_error_position(
                    codeword_block_bits
                )
            )
            corrected_block_bits = list(
                codeword_block_bits
            )

            if error_position > Hamming.CODEWORD_BLOCK_SIZE:
                raise ValueError(
                    "Hamming detectó un síndrome fuera "
                    "del bloque de 12 bits."
                )

            if error_position != 0:
                error_index = error_position - 1
                current_bit = corrected_block_bits[error_index]

                if current_bit == "0":
                    corrected_block_bits[error_index] = "1"
                else:
                    corrected_block_bits[error_index] = "0"

            corrected_blocks.append(
                "".join(corrected_block_bits)
            )

        return "".join(corrected_blocks)

    @staticmethod
    def recover_data(codeword_bits: str) -> str:
        corrected_codeword_bits = Hamming.correct(
            codeword_bits
        )
        recovered_data_bits = []

        for block_start_position in range(
            0,
            len(corrected_codeword_bits),
            Hamming.CODEWORD_BLOCK_SIZE,
        ):
            block_end_position = (
                block_start_position
                + Hamming.CODEWORD_BLOCK_SIZE
            )
            corrected_block_bits = corrected_codeword_bits[
                block_start_position:block_end_position
            ]

            for data_position in Hamming.DATA_POSITIONS:
                data_index = data_position - 1
                recovered_data_bits.append(
                    corrected_block_bits[data_index]
                )

        return "".join(recovered_data_bits)
